File: day08/src/main/python/solution.py
from functools import reduce

from more_itertools import one


def filter_length(lis, length):
    return [x for x in lis if len(x) == length]


class Digit:
    def __init__(self):
        self.top = ' '
        self.middle = ' '
        self.bottom = ' '
        self.topLeft = ' '
        self.topRight = ' '
        self.bottomLeft = ' '
        self.bottomRight = ' '

    def solve(self, description):
        one_value = one(filter_length(description, 2))
        seven = one(filter_length(description, 3))
        four = one(filter_length(description, 4))
        eight = one(filter_length(description, 7))
        self.top = one([d for d in seven if d not in one_value])
        sixers = filter_length(description, 6)
        six = one([x for x in sixers if sum(map(lambda d: d not in one_value, x)) == 5])
        nine = one([x for x in sixers if sum(map(lambda d: d in four, x)) == 4])
        self.topRight = one([d for d in eight if d not in six])
        self.bottomLeft = one([d for d in eight if d not in nine])
        self.bottomRight = one([d for d in one_value if d != self.topRight])
        self.bottom = one([d for d in nine if d not in four and d != self.top])
        zero = one([n for n in sixers if n != six and n != nine])
        self.topLeft = one([d for d in zero if
                            d != self.top and d != self.bottom
                            and d != self.topRight and d != self.bottomRight
                            and d != self.bottomLeft])
        self.middle = one([d for d in four if d != self.topLeft and d != self.topRight and d != self.bottomRight])

    def get_digit(self, line):
        return reduce(lambda acc, d: acc * 10 + self.get_single_digit(d), line, 0)

    def get_single_digit(self, d):
        if len(d) == 2:
            return 1
        elif len(d) == 3:
            return 7
        elif len(d) == 4:
            return 4
        elif len(d) == 5:
            if self.topLeft in d and self.bottomRight in d:
                return 5
            elif self.topRight in d and self.bottomRight in d:
                return 3
            else:
                return 2
        elif len(d) == 6:
            if self.middle not in d:
                return 0
            elif self.bottomLeft not in d:
                return 9
            else:
                return 6
        else:
            return 8


def read_input():
    f = open('../resources/input', 'r')
    descriptions = []
    displays = []
    for line in f.readlines():
        description, display = line.split("|")
        displays.append(display.strip().split(" "))
        descriptions.append(description.strip().split(" "))
    return descriptions, displays


def count_easy_digits(displays):
    return len([x for display in displays for x in display if (2 <= len(x) <= 4) or len(x) == 7])


def count_display_numbers(description, display):
    count = 0
    for i in range(len(description)):
        digit = Digit()
        digit.solve(description[i])
        count += digit.get_digit(display[i])
    return count


if __name__ == "__main__":
    desc, disp = read_input()
    print("part 1")
    print(count_easy_digits(disp))
    print("part 2")
    print(count_display_numbers(desc, disp))

File: day08/src/main/cpp/solution.cpp
#include <algorithm>
#include <cstring>
#include <fstream>
#include <iostream>
#include <iterator>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

#define PVSVS pair<vector<string>, vector<string>>

class Digit {
private:
  char top;
  char middle;
  char bottom;
  char topLeft;
  char topRight;
  char bottomLeft;
  char bottomRight;

public:
  void solve(vector<string> options) {
    string one = *find_if(options.begin(), options.end(),
                          [](string l) { return l.length() == 2; });
    string seven = *find_if(options.begin(), options.end(),
                            [](string l) { return l.length() == 3; });
    string four = *find_if(options.begin(), options.end(),
                           [](string l) { return l.length() == 4; });
    string eight = *find_if(options.begin(), options.end(),
                            [](string l) { return l.length() == 7; });
    top = *find_if(seven.begin(), seven.end(),
                   [one](char l) { return one.find(l) == string::npos; });
    string six = *find_if(options.begin(), options.end(), [one](string s) {
      if (s.length() != 6)
        return false;
      int count = 0;
      for (auto const &f : one) {
        count += s.find(f) == string::npos;
      }
      return count == 1;
    });
    string nine = *find_if(options.begin(), options.end(), [four](string s) {
      if (s.length() != 6)
        return false;
      bool found = true;
      for (auto const &f : four) {
        found &= s.find(f) != string::npos;
      }
      return found;
    });
    topRight = *find_if(eight.begin(), eight.end(),
                        [six](char c) { return six.find(c) == string::npos; });
    bottomLeft = *find_if(eight.begin(), eight.end(), [nine](char c) {
      return nine.find(c) == string::npos;
    });
    bottomRight =
        *find_if(one.begin(), one.end(), [&](char c) { return c != topRight; });
    bottom = *find_if(nine.begin(), nine.end(), [&, four](char c) {
      return four.find(c) == string::npos && c != top;
    });
    string zero =
        *find_if(options.begin(), options.end(), [six, nine](string s) {
          return s.length() == 6 && s != six && s != nine;
        });
    topLeft = *find_if(zero.begin(), zero.end(), [&](char c) {
      return c != top && c != bottom && c != topRight && c != bottomRight &&
             c != bottomLeft;
    });
    middle = *find_if(four.begin(), four.end(), [&](char c) {
      return c != topLeft && c != topRight && c != bottomRight;
    });
  }
  int parseDisplay(vector<string> display) {
    int count = 0;
    for (auto const &digit : display) {
      count = 10 * count + parseDigit(digit);
    }
    return count;
  }

  int parseDigit(string digit) {
    switch (digit.size()) {
    case 2:
      return 1;
    case 3:
      return 7;
    case 4:
      return 4;
    case 5:
      if (digit.find(bottomRight) == string::npos) {
        return 2;
      } else if (digit.find(topLeft) == string::npos) {
        return 3;
      } else {
        return 5;
      }
    case 6:
      if (digit.find(middle) == string::npos) {
        return 0;
      } else if (digit.find(bottomLeft) == string::npos) {
        return 9;
      } else {
        return 6;
      }
    case 7:
      return 8;
    default:
      throw 42;
    }
  }
};

vector<PVSVS> readInputFile() {
  ifstream file("../resources/input");
  vector<PVSVS> result;
  string s;
  while (getline(file, s)) {
    vector<string> digits;
    vector<string> display;
    bool isDigits = true;
    stringstream ss(s);
    for (string word; ss >> word;) {
      if (word == "|") {
        isDigits = false;
        continue;
      }
      if (isDigits) {
        digits.push_back(word);
      } else {
        display.push_back(word);
      }
    }
    result.push_back(pair(digits, display));
  }
  return result;
}

int countSimpleDigits(vector<PVSVS> input) {
  int count = 0;
  for (auto const &value : input) {
    for (auto const &displays : value.second) {
      if ((displays.size() >= 2 && displays.size() <= 4) ||
          displays.size() == 7) {
        count++;
      }
    }
  }
  return count;
}

int countDisplayDigits(vector<PVSVS> input) {
  int count = 0;
  for (auto const &value : input) {
    Digit digit;
    digit.solve(value.first);
    count += digit.parseDisplay(value.second);
  }
  return count;
}

int main() {
  vector<PVSVS> input = readInputFile();
  cout << "part 1" << endl
       << countSimpleDigits(input) << endl
       << "part 2" << endl
       << countDisplayDigits(input) << endl;
  return 0;
}